src/builder.rs

```diff
@@ -0,0 +1,238 @@
+    /// use uuid::UuidBuilder;
+    ///
+    /// let bytes = [4, 54, 67, 12, 43, 2, 98, 76];
+    ///
+    /// let builder = UuidBuilder::from_slice(&bytes);
+    ///
+    /// assert!(builder.is_err());
+    /// ```
+    pub fn from_slice(b: &[u8]) -> Result<Self, BytesError> {
+        const BYTES_LEN: usize = 16;
+
+        let len = b.len();
+
+        if len != BYTES_LEN {
+            return Err(BytesError::new(BYTES_LEN, len));
+        }
+
+        let mut bytes: Bytes = [0; 16];
+        bytes.copy_from_slice(b);
+        Ok(Self::from_bytes_be(bytes))
+    }
+
+    /// Creates a `UuidBuilder` from four field values.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if `d4`'s length is not 8 bytes.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use uuid::UuidBuilder;
+    ///
+    /// let d4 = [12, 3, 9, 56, 54, 43, 8, 9];
+    ///
+    /// let builder = UuidBuilder::from_fields(42, 12, 5, &d4);
+    /// let uuid =
+    ///     builder.map(|builder| builder.build().to_hyphenated().to_string());
+    ///
+    /// let expected_uuid =
+    ///     Ok(String::from("0000002a-000c-0005-0c03-0938362b0809"));
+    ///
+    /// assert_eq!(expected_uuid, uuid);
+    /// ```
+    ///
+    /// An invalid length:
+    ///
+    /// ```
+    /// use uuid::prelude::*;
+    ///
+    /// let d4 = [12];
+    ///
+    /// let builder = uuid::UuidBuilder::from_fields(42, 12, 5, &d4);
+    ///
+    /// assert!(builder.is_err());
+    /// ```
+    pub fn from_fields(
+        d1: u32,
+        d2: u16,
+        d3: u16,
+        d4: &[u8],
+    ) -> Result<Self, BytesError> {
+        Uuid::from_fields(d1, d2, d3, d4).map(|uuid| UuidBuilder { uuid })
+    }
+
+    /// Creates a `UuidBuilder` with an initial [`Uuid::nil`]
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use uuid::UuidBuilder;
+    ///
+    /// let builder = UuidBuilder::nil();
+    ///
+    /// assert_eq!(
+    ///     builder.build().to_hyphenated().to_string(),
+    ///     "00000000-0000-0000-0000-000000000000"
+    /// );
+    /// ```
+    pub fn nil() -> Self {
+        UuidBuilder { uuid: Uuid::nil() }
+    }
+
+    /// Specifies the variant of the internal [`Uuid`].
+    pub fn set_variant(&mut self, v: Variant) -> &mut Self {
+        self.uuid.set_variant(v);
+        self
+    }
+
+    /// Specifies the version number of the internal [`Uuid`].
+    pub fn set_version(&mut self, v: Version) -> &mut Self {
+        self.uuid.set_version(v);
+        self
+    }
+
+    /// Hands over the internal constructed [`Uuid`]
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use uuid::UuidBuilder;
+    ///
+    /// let uuid = UuidBuilder::nil().build();
+    ///
+    /// assert_eq!(
+    ///     uuid.to_hyphenated().to_string(),
+    ///     "00000000-0000-0000-0000-000000000000"
+    /// );
+    /// ```
+    pub fn build(&self) -> Uuid {
```

Comment:
`pub fn build(&self) -> Uuid` -> `pub fn build(self) -> Uuid`

Reply:
We could make this method take `&mut self`, so that we have the future option to move the internal uuid out instead of copying it.

Reply:
I used `&self` since that was what was given in the issue example, normally you'd consume the builder when building but since our internal type is copy both works here I'd say. What should I pick?

Reply:
I would go with `&mut self` here, it's consistent with the other mutating methods we've added, and lets us do some extra work in the `build` method if we need it. Using `&mut self` over `self` means that we can chain the method calls in a single statement like this:

```rust
let uuid = Builder::from_slice(bytes).set_version(..).build();
```

If the `build` method took `self` by value then we wouldn't be able to chain it with `set_version`.